bids/layout/index.py

```diff
@@ -208,43 +210,37 @@ def _index_dir(self, path, config, default_action=None):
         for c in config:
             config_entities.update(c.entities)
 
-        # Get a list of first-level subdirectories and files in the path directory
+        # Get lists of 1st-level subdirectories and files in the path directory
         _, dirnames, filenames = next(os.walk(path))
 
-        # Set the default inclusion/exclusion directive
-        default = _validate_path(
-            path,
-            incl_patt=self._include_patterns,
-            excl_patt=self._exclude_patterns,
-            default=default_action,
-            root=self._layout._root,
-        )
-
         # If layout configuration file exists, delete it
         if self.config_filename in filenames:
             filenames.remove(self.config_filename)
 
         for f in filenames:
-            bf = self._index_file(f, path, config_entities,
-                                  default_action=default)
-            if bf is None:
+            abs_fn = path / f
+            # Skip files that fail validation, unless forcibly indexing
+            if not self._validate_file(abs_fn):
                 continue
 
+            self._index_file(abs_fn, config_entities)
+
         self.session.commit()
 
         # Recursively index subdirectories
         for d in dirnames:
             d = path / d
-            self._index_dir(d, list(config), default_action=default)
+            valid_dir = _validate_path(
+                d,
+                incl_patt=self._include_patterns,
+                excl_patt=self._exclude_patterns,
+                root=self._layout._root,
+            )
+            if valid_dir is not False:
+                self._index_dir(d, list(config), default_action=default_action)
```

Comment:
`default_action` no longer has any effect at all. It's just a bool passed recursively to `_index_dir()`.

If I understand it correctly, it was supposed to allow a directory to index everything underneath it with `force_index`. But it was getting set on any valid path because `_validate_path()` returned `True`.  I think instead we want to make this a `force=` option that only gets set if there's a force_index match, not if there's a valid path. Given that it's extremely rare for someone to force index, we would preserve the inefficient behavior for them while correctly pruning for the majority of users.

Reply:
Happy to see the alternative :)

Reply:
Do you have a benchmark dataset that you'd like to make sure doesn't regress under my proposal?

Reply:
BTW, we have discussed dropping `force_index` entirely in the upcoming changes. It's a very niche situation we probably should have never added.